### DISCUSSION TITLE
smallrye-reactive-messaging-jms with ActiveMQ Classic - failure handling/DLQ

### DISCUSSION THREAD

**Discussion by tobi390** [smallrye/smallrye-reactive-messaging#2439] [Q&A]
I tested quarkus-smallrye-reactive-messaging-amqp with ActiveMQ Classic 5.18.3 and server side redelivery configuration successfull. In case of failure the messages go to my DLQ. Quarkus version 3.2.9.Final.

Now I'd like to test the same with smallrye-reactive-messaging-jms (ActiveMQ Classic configuration is the same like in AMQP test except that AMQP is deactivated and openwire is configured).

My config:
```
mp.messaging.connector.smallrye-jms.connection-factory-name=ActiveMQClassicConnectionFactory
mp.messaging.incoming.requests.connector=smallrye-jms
mp.messaging.incoming.requests.host=activemq
mp.messaging.incoming.requests.session-mode=AUTO_ACKNOWLEDGE

mp.messaging.outgoing.quotes.connector=smallrye-jms
mp.messaging.outgoing.quotes.host=activemq
```
My Method:
```
    @Incoming("requests") 
    @Outgoing("quotes") 
    @Transactional(Transactional.TxType.REQUIRED)
    public Quote process(String quoteRequest){
        if ((System.currentTimeMillis() / 1000) % 3 != 0) {
            Log.info("Failed to consume the message ");
            throw new RuntimeException("Failed to consume");
        }

        Log.info("Successfully consumed message");
        return new Quote(quoteRequest, random.nextInt(100));
    }
```
My bean producer method (activemq-client:6.0.0):
```
	@Produces
	@Typed({ ConnectionFactory.class })
	@Named("ActiveMQClassicConnectionFactory")
	@ApplicationScoped
	ActiveMQConnectionFactory create() {
		ActiveMQConnectionFactory factory = new ActiveMQConnectionFactory();
		factory.setBrokerURL("tcp://activemq:61616");

		return factory;
	}
```

In case the RuntimeException is thrown, the message just gets acknowledged (instead of pushed back and retried) and no message is sent to the outgoing queue.
In case no exception is thrown, the message passes the queue successfully.

I did not find much documentation on smallrye-reactive-messaging-jms and error handling. Is it possible to push back messages to the queue, redeliver them and (in case of another Exception) get them to DLQ? 

I miss some config like in AMQP where I could add some failure-strategy: failure-strategy=modified-failed-undeliverable-here

**Comment by Bidek56** [comment]
In [ActiveMQ](https://activemq.apache.org/components/artemis/documentation/latest/undelivered-messages), for a message to end up in DLQ, it needs to be rolled back.
Do not throw `RuntimeException` it's considered a programming error.

**Reply by tobi390** [reply]
Thank you for your reply. 
I had to take some time off because my son had to be introduced to a day-care facility, but now I'm back on the topic.

I changed the initial process method to:
```
    @Incoming("requests") 
    @Transactional(Transactional.TxType.REQUIRED)
    public CompletionStage<Void> process(Message<String> quoteMessage){
        try {
            if ((System.currentTimeMillis() / 1000) % 3 != 0) {
                Log.info("Failed to consume the message ");
                throw new Exception("Failed to consume");
            }
  
            Log.info("Successfully consumed message");
            return quoteMessage.ack()
        catch (Exception e) {
            QuarkusTransaction.setRollbackOnly();
            return quoteMessage.nack(e);
        }
    }
```
Quarkus was updated to 3.8.4, smallrye-reactive-messaging-jms to 4.21.0.

My expectation was, now that we have a clear nack on the message, that the message would be retried and end up in DLQ. But the behaviour did not change, the message is still acknowledged in the error case.

Mabe I did not get your comment right, @Bidek56 ?

Please be aware that ActiveMQ Classic 5.18.3 is used, not ActiveMQ Artemis.
